### ISSUE TITLE
Support Root-Relative Links (Absolute Links)?

### ISSUE THREAD

**Issue by karlhorky** [remarkjs/remark-validate-links#57]
Hi @wooorm ! 👋

### Subject of the feature

What do you think about supporting root-relative links such as in the example below?

```md
[abc](/docs/non-existent/)
[abc](/docs/exists/#non-existent2)
```

I suppose this could be configured with a configuration option, maybe named like `basedir`.

### Problem

On large sites with a lot of moving files or large directory structures, updating relative links can be a pain.

Root relative links can also avoid any ambiguity that can be caused by using the same path ending and file name between multiple files.

### Expected behaviour

It would be nice for these links to also be checked.

### Alternatives

[gatsby-remark-check-links](https://www.npmjs.com/package/gatsby-remark-check-links), but:

1. It's Gatsby-specific
2. It runs as a Gatsby plugin and only warns in the console :(

[gh-action-check-broken-links](https://github.com/hashicorp/gh-action-check-broken-links), but:

1. It's wrapped in a GitHub Action
2. It assumes everything is in a `pages` directory

**Comment by wooorm** [comment]
Say I have a `package.json` like this:

```js
{
  "name": "example",
  "scripts": {
    "test": "remark ."
  },
  "remarkConfig": {
    "plugins": [
      [
        "validate-links",
        {
          "repo": "x/y"
        }
      ]
    ]
  },
  "devDependencies": {
    "remark-cli": "^8.0.1",
    "remark-validate-links": "^10.0.2"
  }
}
```

A `readme.md` file:

```md
[a](/x/y/blob/main/docs/non-existent/)
[b](/x/y/blob/main/docs/exists/#readme)
[c](/x/y/blob/main/docs/exists/#non-existent2)
```

And a `docs/exists/readme.md` file:

```md
Hi!
```

Now, running `npm test` yields:

```
$ nr test

> example@ test /Users/tilde/Projects/oss/example
> remark .

docs/exists/readme.md: no issues found
readme.md
  1:1-1:39  warning  Link to unknown file: `docs/non-existent`                  missing-file             remark-validate-links
  3:1-3:47  warning  Link to unknown heading in `docs/exists`: `non-existent2`  missing-heading-in-file  remark-validate-links

⚠ 2 warnings
```

As you can see, the links are checked. `remark-validate-links` correctly sees that `non-existent` does not exist, that `docs/exists` is a folder with a `readme.md`, but that it does not have a `#non-existent2` header.

This is to say, we do support full urls with a hostname (`https://github.com/x/y/blob/main/docs/exists`), absolute paths  (`/x/y/blob/main/docs/exists`), relative paths and whatnot too.
What this project does is resolve all URLs, as if the markdown was rendered on GH, Gl, Bb. It checks whether they would work or not.

---

What you mention here:

> ...along with this link in Markdown...
> ```
> [link](/remarkjs/remark-validate-links/docs/configuration)
> ```
>...would search for the following files in the repo to see if it will be a broken link:
> ```
> /docs/configuration.md
> /docs/configuration/index.md
> ```
> This would allow for a lot of cases for Markdown projects, both:

That sort of works as shown above, but it does that with `readme.md` files. (And, `blob/branch` is needed).

---

So, I’d say the question, support root-relative links (absolute links), is already solved. Just in a different way than what you expected. Thus, to come back to the original problem

> On large sites with a lot of moving files or large directory structures, updating relative links can be a pain.
>
>Root relative links can also avoid any ambiguity that can be caused by using the same path ending and file name between multiple files.

Can you tell me more about this?

**Comment by karlhorky** [comment]
Sure. Here's an example:

1. A user has a website or blog built with MD/MDX files in a `src/pages` directory (eg. on Gatsby, Next.js, 11ty, etc)
2. The MD/MDX files link amongst themselves. These links will become links on the built website:
   ```md
   To read more, check out [the docs](/lib-2/docs) about `lib-2`!
   ```
   This link above would link to the built result of the file `/src/pages/lib-2/docs`. The configuration here would probably look something like: `baseDir: '/src/pages'`
3. Optional (to illustrate ambiguity): This website or blog has a structure that is repeated across multiple pages (eg. a `docs.md` file or `docs` folder inside many different folders)
4. Optional (to illustrate files / directories that move): This is a fast-moving project, so the structure of the site ends up changing every 6 months to 2 years.
5. Users want to be able to be certain that their links work and continue to work

Does this clarify anything?

**Comment by karlhorky** [comment]
I ended up building this in the meantime:

https://github.com/upleveled/mdx-local-link-checker

**Comment by karlhorky** [comment]
So in case [`urlConfig`](https://github.com/remarkjs/remark-validate-links#urlconfig) or some other configuration setting would be changed in `remark-validate-links` (unlikely, given that this project is GitHub / GitLab / Bitbucket focused), maybe it could work like this:

Config:

```js
baseDir: '/src/pages',
extension: 'mdx',
```

`src/pages/lib-1/index.mdx`

```mdx
To read more, check out [the docs](/lib-2/docs) about `lib-2`!
```

The link above would be resolved to the file `<repository root>/src/pages/lib-2/docs.mdx`

**Comment by LuchoTurtle** [comment]
Having this feature would be extremely useful, especially in monorepos where some markdowns reference other projects. Any news on this? 👀

**Comment by wooorm** [comment]
Absolute paths are now supported like GitHub again, per https://github.com/remarkjs/remark-validate-links/issues/75.
For arbitrary websites, so not markdown files on GitHub/Gitlab/Bitbucket, use [`remark-lint-no-dead-urls`](https://github.com/remarkjs/remark-lint-no-dead-urls)
